libgadget/blackhole.c

```diff
@@ -185,8 +205,61 @@ static double blackhole_soundspeed(double entropy, double rho) {
     return cs;
 }
 
+static void
+collect_BH_info(int * ActiveParticle,int NumActiveParticle, struct BHPriv *priv, FILE * FdBlackholeDetails)
+{
+    int i;
+    int c=0;
+    
+    for(i = 0; i < NumActiveParticle; i++)
+    {
+        int p_i = ActiveParticle ? ActiveParticle[i] : i;
+        
+        if(P[p_i].Type != 5 || P[p_i].IsGarbage || P[p_i].Mass <= 0)
+          continue;
+        
+        int PI = P[p_i].PI;
+        
+        struct BHinfo info;
+        info.ID = P[p_i].ID;
+        info.Mass = BHP(p_i).Mass;
+        info.Mdot = BHP(p_i).Mdot;
+        info.Density = BHP(p_i).Density;
+        info.minTimeBin = BHP(p_i).minTimeBin;
+        info.MinPotPos[0] = BHP(p_i).MinPotPos[0];
+        info.MinPotPos[1] = BHP(p_i).MinPotPos[1];
+        info.MinPotPos[2] = BHP(p_i).MinPotPos[2];
+        
+        info.MinPot = priv->MinPot[PI];
+        info.BH_Entropy = priv->BH_Entropy[PI];
+        info.BH_SurroundingGasVel[0] = priv->BH_SurroundingGasVel[PI][0];
+        info.BH_SurroundingGasVel[1] = priv->BH_SurroundingGasVel[PI][1];
+        info.BH_SurroundingGasVel[2] = priv->BH_SurroundingGasVel[PI][2];
+        
+        info.BH_accreted_BHMass = priv->BH_accreted_BHMass[PI];
+        info.BH_accreted_Mass = priv->BH_accreted_Mass[PI];
+        info.SPH_SwallowID = priv->SPH_SwallowID[PI];
+        
+        info.a = All.Time;
+        
+        int size = sizeof(info);
+        
+        fwrite(&size, sizeof(size), 1, FdBlackholeDetails);
+        fwrite(&info,sizeof(info),1,FdBlackholeDetails);
+        fwrite(&size, sizeof(size), 1, FdBlackholeDetails);
+        c++;
+    }
+    
+    fflush(FdBlackholeDetails);
```

Comment:
The fflush is overkill for debug information and will probably just slow us down.

Reply:
I run two sets with BlackHoleWriteDetails on and off. According to cpu.txt, the blackhole.c constitutes 4.9% when turning the flag on, and 4.5% with the flag off.  
The writing procedure takes about 0.2s for a PM-step (Nf ~ 1e9) and about 0.01s for a typical (Nf ~ 1e5) time step.  I would say it doesn't have a significant time impact?

Reply:
That is a convincing answer - but if you could remove the fflush anyway I'd appreciate it. Sometimes doing this kind of thing puts extra stress on the filesystems and the cluster breaks.

Reply:
Sure.  I am a bit surprised though, BlueTides (BT2 and BT3 I mean) also output BH details at every time step with this fflush(FdBlackholeDetail), ---- and it was doing the output in 20K MPI ranks!  I hope it didn't account for the node failures I met in BT runs, otherwise I would regret that I didn't know this before...

Reply:
Eh, well, if it worked on BT just leave it in. I'm probably just getting paranoid because frontera has been so flaky.